cli/lsp/repl.rs

```diff
@@ -0,0 +1,308 @@
+    })
+    .await
+  }
+
+  pub async fn completions(
+    &mut self,
+    line_text: &str,
+    position: usize,
+  ) -> Vec<ReplCompletionItem> {
+    with_logging_disabled(async {
+      self.did_change(line_text).await;
+      let before_line_len = BytePos(self.document_text.len() as u32);
+      let position = before_line_len + BytePos(position as u32);
+      let text_info = deno_ast::SourceTextInfo::from_string(format!(
+        "{}{}",
+        self.document_text, self.pending_text
+      ));
+      let line_and_column = text_info.line_and_column_index(position);
+      let response = self
+        .language_server
+        .completion(CompletionParams {
+          text_document_position: TextDocumentPositionParams {
+            text_document: TextDocumentIdentifier {
+              uri: self.get_document_specifier(),
+            },
+            position: Position {
+              line: line_and_column.line_index as u32,
+              character: line_and_column.column_index as u32,
+            },
+          },
+          work_done_progress_params: WorkDoneProgressParams {
+            work_done_token: None,
+          },
+          partial_result_params: PartialResultParams {
+            partial_result_token: None,
+          },
+          context: None,
+        })
+        .await
+        .ok()
+        .unwrap_or_default();
+
+      let items = match response {
+        Some(CompletionResponse::Array(items)) => items,
+        Some(CompletionResponse::List(list)) => list.items,
+        None => Vec::new(),
+      };
+      items
+        .into_iter()
+        .filter_map(|item| {
+          item.text_edit.and_then(|edit| match edit {
+            CompletionTextEdit::Edit(edit) => Some(ReplCompletionItem {
+              new_text: edit.new_text,
+              span: lsp_range_to_span(&text_info, &edit.range),
+            }),
+            CompletionTextEdit::InsertAndReplace(_) => None,
+          })
+        })
+        .filter(|item| {
+          // filter the results to only exact matches
+          let text = &text_info.text_str()
+            [item.span.lo.0 as usize..item.span.hi.0 as usize];
+          item.new_text.starts_with(text)
+        })
+        .map(|mut item| {
+          // convert back to a line position
+          item.span = Span::new(
+            item.span.lo - before_line_len,
+            item.span.hi - before_line_len,
+            Default::default(),
+          );
+          item
+        })
+        .collect()
+    })
+    .await
+  }
+
+  async fn did_change(&mut self, new_text: &str) {
+    self.check_cwd_change().await;
+    let new_text = if new_text.ends_with('\n') {
+      new_text.to_string()
+    } else {
+      format!("{}\n", new_text)
+    };
+    self.document_version += 1;
+    let current_line_count =
+      self.document_text.chars().filter(|c| *c == '\n').count() as u32;
+    let pending_line_count =
+      self.pending_text.chars().filter(|c| *c == '\n').count() as u32;
```

Comment:
Are these related to the current line of input or the full REPL history? In the latter case O(n) scans might turn problematic for long-lived sessions.

Reply:
Full history for the current session probably isn't so bad.